design/baremetal-operator/bmc-events.md

```diff
@@ -0,0 +1,143 @@
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Event Subscription API
+
+Users of bare metal hardware may want to receive events from the
+baseboard management controller (BMC) in order to act on them in the
+event of a hardware fault, increase in temperature, removal of a
+device, etc.
+
+## Status
+
+provisional
+
+## Summary
+
+The Redfish standard includes the ability to subscribe to events, which
+will cause hardware events to be sent in a particular format to a destination
+URI. This design document describes a Metal3 API for configuring a
+subscription. While Redfish is the primary target for this design, the
+Ironic API is vendor-neutral and seeks to provide a unified interface
+for configuring events.
+
+## Motivation
+
+Some environments run workloads that need to deal with potential
+faults or environmental changes quicker than they would get an alert
+through other channels. For example, some workloads may have a sidecar
+container that knows how to deal with an alert that a particular network
+interface went down, or that the CPU temperature reached a certain
+threshold.
+
+### Goals
+
+- Provide an API to manage subscriptions to events
+
+### Non-Goals
+
+- [Configurable events and thresholds](#configurable-events-and-thresholds)
+- Any kind of event polling
+- Software for processing the events, i.e. any webhook
+- BMC's beyond Redfish for now
+
+## Proposal
+
+### User Stories
+
+- I'd like to configure my BMC to send events to a destination URI.
+- I'd like to provide context to a particular event subscription.
+- I'd like to provide arbitrary HTTP headers
+- I'd like the baremetal-operator to reconcile on the
+  BMCEventSubscription resource, and ensure its state is accurate in
+  Ironic.
+
+## Design Details
+
+### Implementation Details
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BMCEventSubscription
+metadata:
+  name: worker-1-events
+spec:
+   hostRef: ostest-worker-1
+   destination: https://events.apps.corp.example.com/webhook
+   context: “SomeUserContext”
+   httpHeadersRef:
+     name: some-secret-name
+     namespace: some-namespace
+status:
+  errorMessage: ""
+  subscriptionID: aa618a32-9335-42bc-a04b-20ddeed13ade
+```
+
+- A BMCEventSubscription resource represents a subscription to the events generated
+  by a specific BMC.
+- Ironic will manage configuring the subscription using a vendor passthru API.
+- The BMCEventSubscription will maintain a reference to a BareMetalHost.
+- The BMCEventSubscription will allow injection of headers using a
+  reference to a secret, for example to provide basic auth credentials.
+- The BMCEventSubscription will reside in the same namespace as the referenced
+  BareMetalHost.
+- The BMCEventSubscription will maintain a reference to the subscription ID
+  obtained from the BMC.
+- The baremetal-operator binary will be expanded to include an additional
+  reconciler with a dedicated controller/reconcile loop for
+  BMCEventSubscriptions.
+
```

Comment:
Is it possible to expand this section to define more precisely the expected behavior in case of host deletion? In case the subscription resource remains alive, what will happen when/if the host will come back? What will be status shown to the user meanwhile?